File: src/labeling/rules.py
"""rules.py — motor de reglas deterministas para la determinación deóntica
(Von Wright) y la posición hohfeldiana por defecto.

Distingue dos niveles de confianza, documentados en `notes` cuando aplica:

  * marcador léxico de baja ambigüedad medido o evidenciado dentro del
    propio proyecto — antecedent_operator, exception,
    presuncion/remision/definicion, deóntica de Von Wright.
  * sin marcador textual, pero con un valor modal razonable y documentado
    en `notes` — p. ej. addressee="partes" para una regla con deóntica
    explícita (la mayoría del derecho privado se dirige a las partes salvo
    marca en contrario), o el correlato hohfeldiano por defecto de la
    deóntica detectada (ver `_derive_hohfeld_from_deontic`: es una
    simplificación declarada, NO un análisis bilateral completo de
    Hohfeld, que exigiría identificar a la contraparte concreta).

Cuando no hay evidencia suficiente, el campo queda en None (o su valor
"ninguno"/"ninguno_explicito"), nunca adivinado. No usa red, no usa LLM, no
tiene costo.
"""

from __future__ import annotations

import re
import unicodedata

from src.labeling import lexical_markers as lex
from src.labeling.schemas import LabelProposal, ProlegPreview, ProlegRunResult
from src.models import COMPATIBILITY
from src.reasoning.engine import prove
from src.reasoning.models import FactAction, FactBase, FactEntry, Party, Rule, RuleBase, ExceptionLink


# Artefacto de la extracción del Código Civil brasileño (152 artículos): la
# letra inicial de un párrafo quedó tachada y separada del resto de la
# palabra — "~~N~~ ão pode", "~~S~~ alvo quando". Sin volver a unirla, el
# motor no ve la negación ni la excepción, y "Não pode o devedor…" se leía
# como un permiso. Solo se une cuando lo que sigue es una letra, es decir,
# cuando es claramente la misma palabra.
_LETRA_INICIAL_TACHADA = re.compile(r"~~([^\W\d_])~~\s+(?=[^\W\d_])")


def _detect_antecedent_operator(text: str) -> str:
    for label, pattern in lex.ANTECEDENT_PATTERNS:
        if pattern.search(text):
            return label
    return "ninguno_explicito"


def _detect_exception(text: str) -> tuple[bool, str | None, str | None]:
    for marker_label, pattern in lex.EXCEPTION_PATTERNS:
        match = pattern.search(text)
        if match:
            tail = text[match.start() : match.start() + 300]
            scope = "por_remision" if lex.ARTICLE_REFERENCE_RE.search(tail) else "interna"
            return True, marker_label, scope
    return False, None, None


def _detect_presumption(text: str) -> tuple[bool, bool | None]:
    if lex.PRESUMPTION_DE_DERECHO_RE.search(text):
        return True, False
    if lex.PRESUMPTION_LEGAL_RE.search(text):
        return True, True
    return False, None


def _detect_remision(text: str) -> bool:
    n_words = len(text.split())
    return n_words <= lex.REMISION_MAX_WORDS and bool(lex.REMISION_LEAD_RE.search(text))


def _detect_definicion(text: str) -> bool:
    return bool(lex.DEFINICION_LEAD_RE.search(text))


def _detect_deontic_modality(text: str) -> str | None:
    """Von Wright: obligación / prohibición / permiso. Orden importa —
    prohibición antes que permiso ("no podrá" no debe matchear "podrá")."""
    if lex.DEONTIC_PROHIBICION_RE.search(text):
        return "prohibicion"
    if lex.DEONTIC_OBLIGACION_RE.search(text):
        return "obligacion"
    if lex.DEONTIC_PERMISO_RE.search(text):
        return "permiso"
    return None


def _derive_hohfeld_from_deontic(deontic_modality: str | None, addressee: str | None) -> str:
    """Correlato hohfeldiano por defecto de la modalidad deóntica detectada.

    Simplificación declarada: Hohfeld exige identificar a la CONTRAPARTE
    concreta (quien tiene el deber correlativo del derecho, o quien está
    sujeto al ejercicio de la potestad) — algo que un texto
    normativo aislado no siempre da. Esta derivación asigna la posición del
    DESTINATARIO del mandato (no de su contraparte), que es lo único que el
    propio enunciado permite fijar sin inventar quién más interviene:
    obligación/prohibición -> deber (el destinatario debe/no debe actuar).

    "permiso" se bifurca porque colapsa dos casillas hohfeldianas DISTINTAS
    (no una simplificación de una sola): una LIBERTAD/PRIVILEGIO (facultad
    de actuar sin alterar la posición jurídica de nadie más — "podrá
    dedicarse libremente a un empleo") no es lo mismo que una POTESTAD
    (capacidad de alterar unilateralmente relaciones jurídicas ajenas — "el
    juez podrá reducir la pena"). Se usa el destinatario ya detectado para
    distinguirlas: permiso dirigido a juez/funcionario -> potestad (altera
    la posición de otro); permiso dirigido a partes/tercero/sin marcador ->
    privilegio (libertad civil ordinaria, el caso mayoritario).
    Ver docs/limitaciones_conocidas.md §2 para el límite de fondo (falta de
    correlatividad bilateral, un problema distinto y adicional a este)."""
    if deontic_modality in ("obligacion", "prohibicion"):
        return "deber"
    if deontic_modality == "permiso":
        if addressee in ("juez", "funcionario_o_notario"):
            return "potestad"
        return "privilegio"
    return "ninguno"


def _detect_addressee(text: str) -> str | None:
    if lex.ADDRESSEE_JUEZ_RE.search(text):
        return "juez"
    if lex.ADDRESSEE_FUNCIONARIO_RE.search(text):
        return "funcionario_o_notario"
    if lex.ADDRESSEE_TERCERO_RE.search(text):
        return "tercero"
    return None


def _detect_enumeration(text: str) -> tuple[bool, bool | None]:
    if not lex.ENUMERATION_ITEM_RE.search(text):
        return False, None
    if lex.ENUMERATION_CLOSED_RE.search(text):
        return True, True
    if lex.ENUMERATION_OPEN_RE.search(text):
        return True, False
    return True, None  # hay lista, pero abierta/cerrada queda sin determinar


def _derive_structure(
    statement_type: str | None, exception_present: bool, has_enumeration: bool
) -> str | None:
    """Usa COMPATIBILITY para derivar structure solo cuando queda un único
    candidato posible — nunca elige entre varios igual de válidos."""
    if statement_type is None:
        return None
    candidates = COMPATIBILITY.get(statement_type, set())
    if not candidates:
        return None
    if exception_present:
        if "supuesto_consecuencia_con_excepcion" in candidates:
            return "supuesto_consecuencia_con_excepcion"
        return None
    remaining = candidates - {"supuesto_consecuencia_con_excepcion"}
    if not has_enumeration:
        remaining = remaining - {"enumeracion_taxativa", "enumeracion_enunciativa"}
    if len(remaining) == 1:
        return next(iter(remaining))
    return None


def build_proleg_preview(
    exception_marker: str | None, exception_scope: str | None, statement_type: str | None = None
) -> ProlegPreview:
    """Instancia una RuleBase mínima y GENÉRICA a partir de la MISMA
    estructura ya detectada (regla + excepción) y corre el motor PROLEG
    real dos veces: una sin la excepción probada, otra con ella. No fabrica
    contenido semántico del artículo — los nombres de hecho
    ("antecedente_cumplido", "excepcion_probada") son genéricos a
    propósito, para no simular una prueba jurídica que el texto por sí solo
    no permite construir. Lo que demuestra es la MECÁNICA de
    derrotabilidad, con el motor real, no un mock.

    `statement_type` cambia el TEXTO de la nota, no el cálculo: en una
    presunción, la "excepción" que el motor modela es en realidad la
    prueba en contrario, que opera por desplazamiento de la CARGA
    PROBATORIA (quien quiere desvirtuar la presunción debe probarlo) — una
    figura procesal distinta de la excepción sustantiva ordinaria de una
    regla (un hecho impeditivo que, probado, derrota la consecuencia). El
    motor PROLEG usa el mismo mecanismo de regla+excepción para ambas
    porque estructuralmente se comportan igual (algo se prueba salvo que
    se pruebe lo contrario), pero la nota debe decir cuál de las dos
    figuras jurídicas está mostrando, para no presentarlas como si fueran
    la misma cosa."""
    rulebase = RuleBase(
        id="preview_derrotabilidad",
        description=(
            f"Estructura genérica derivada del artículo: regla con excepción "
            f'("{exception_marker}", scope={exception_scope}).'
        ),
        rules=[
            Rule(
                head="consecuencia_aplica",
                body=["antecedente_cumplido"],
                source_note="Regla base detectada por reglas léxicas (no contenido inventado).",
            )
        ],
        exceptions=[ExceptionLink(rule_head="consecuencia_aplica", exception_head="excepcion_probada")],
    )

    base_facts = FactBase(
        entries=[
            FactEntry(action=FactAction.ALLEGE, fact="antecedente_cumplido", party=Party.PLAINTIFF),
            FactEntry(action=FactAction.PROVIDE_EVIDENCE, fact="antecedente_cumplido", party=Party.PLAINTIFF),
            FactEntry(action=FactAction.PLAUSIBLE, fact="antecedente_cumplido", party=None),
        ]
    )

    without_exception = prove("consecuencia_aplica", Party.PLAINTIFF, rulebase, base_facts)

    with_exception_facts = FactBase(
        entries=list(base_facts.entries)
        + [FactEntry(action=FactAction.PLAUSIBLE, fact="excepcion_probada", party=None)]
    )
    with_exception = prove("consecuencia_aplica", Party.PLAINTIFF, rulebase, with_exception_facts)

    if statement_type == "presuncion":
        note = (
            "Vista previa estructural: usa el motor de razonamiento real "
            "sobre una regla genérica con la misma forma detectada en el "
            "artículo. En una presunción, esto modela la PRUEBA EN "
            "CONTRARIO (desplazamiento de la carga de la prueba hacia "
            "quien quiere desvirtuarla), no la derrota de una regla "
            "sustantiva — son figuras procesales distintas aunque el "
            "motor las calcule con el mismo mecanismo. No es un análisis "
            "semántico del contenido específico del artículo."
        )
    else:
        note = (
            "Vista previa estructural: usa el motor de razonamiento real "
            "sobre una regla genérica con la misma forma detectada en el "
            "artículo (regla + excepción sustantiva). No es un análisis "
            "semántico del contenido específico del artículo — muestra "
            "que la excepción detectada, si se prueba, efectivamente "
            "derrota la regla bajo el motor determinista."
        )

    return ProlegPreview(
        rulebase_id=rulebase.id,
        without_exception=ProlegRunResult(
            proved=without_exception.proved,
            trace_length=len(without_exception.trace),
        ),
        with_exception=ProlegRunResult(
            proved=with_exception.proved,
            trace_length=len(with_exception.trace),
        ),
        note=note,
    )


def propose_from_text(text: str) -> LabelProposal:
    # NFC: sin esto, el mismo texto en dos formas Unicode canónicamente
    # equivalentes (p. ej. copiado desde macOS o extraído de OCR, que suelen
    # producir NFD) puede no matchear los patrones léxicos (que usan tildes
    # precompuestas) y dar un resultado distinto para el mismo enunciado —
    # rompería la promesa de determinismo de este motor. El pipeline de
    # ingesta (src/pipeline.py) ya normaliza a NFC los 8 corpus reales; esto
    # cubre el otro punto de entrada: texto pegado directo por un usuario.
    text = unicodedata.normalize("NFC", text).strip()
    text = _LETRA_INICIAL_TACHADA.sub(r"\1", text)

    notes: list[str] = []

    antecedent_operator = _detect_antecedent_operator(text)

    exception_present, exception_marker, exception_scope = _detect_exception(text)

    deontic_modality = _detect_deontic_modality(text)
    deontic_determined = deontic_modality is not None
    if not deontic_determined:
        deontic_modality = "ninguno"
        notes.append(
            "Deóntica (Von Wright): sin marcador léxico de obligación/"
            "prohibición/permiso — se deja 'ninguno' por defecto (correcto "
            "para definiciones/remisiones, pero revisar si es una regla "
            "con deóntica implícita en presente indicativo, ej. 'el "
            "comprador paga el precio' sin 'deberá')."
        )

    # addressee se detecta ACÁ (antes de derivar Hohfeld) porque
    # _derive_hohfeld_from_deontic necesita saber si el "permiso" se
    # dirige a un juez/funcionario (potestad) o a partes/tercero
    # (privilegio) — ver docstring de esa función.
    addressee = _detect_addressee(text)
    if addressee is None and deontic_modality != "ninguno":
        addressee = "partes"
        notes.append(
            "Destinatario: sin marcador de juez/funcionario/tercero — se "
            "asume 'partes' por defecto (mayoría del derecho privado); "
            "revisar si el artículo se dirige a otro destinatario."
        )

    # "tiene derecho a" fija Hohfeld POR SÍ MISMO, sin pasar por Von Wright:
    # no manda una conducta, afirma una posición jurídica cuyo correlativo es
    # un deber en la otra parte. Se chequea antes que la derivación desde la
    # deóntica porque es evidencia más directa que un correlato por defecto.
    if lex.DERECHO_SUBJETIVO_RE.search(text):
        hohfeldian_position = "derecho_subjetivo"
        notes.append(
            "Posición (Hohfeld): el texto dice 'tiene derecho a', que afirma "
            "un derecho subjetivo — su correlativo es un deber en la otra "
            "parte, no la mera libertad de un privilegio. No se deriva de la "
            "deóntica; se lee directo del enunciado."
        )
    else:
        hohfeldian_position = _derive_hohfeld_from_deontic(
            deontic_modality if deontic_determined else None,
            addressee,
        )
        if deontic_determined:
            notes.append(
                "Posición (Hohfeld): correlato por defecto de la deóntica "
                "detectada (y, si es 'permiso', del destinatario) — no un "
                "análisis bilateral de Hohfeld (no identifica contraparte) — "
                "ver docs/limitaciones_conocidas.md §2."
            )

    statement_type: str | None = None
    presumption_rebuttable: bool | None = None

    is_presumption, rebuttable = _detect_presumption(text)
    if is_presumption:
        statement_type = "presuncion"
        presumption_rebuttable = rebuttable
    elif _detect_remision(text):
        statement_type = "remision"
    elif _detect_definicion(text):
        statement_type = "definicion"
    elif deontic_modality != "ninguno":
        # Hay marcador deóntico explícito y ningún otro marcador más
        # específico (presunción/remisión/definición): el default modal es
        # "regla" — es, literalmente, la definición de statement_type=regla
        # en este esquema (mandato con antecedente y consecuente).
        statement_type = "regla"
    else:
        notes.append(
            "Tipo de norma: no determinado. Sin marcador léxico de "
            "presunción/remisión/definición ni deóntica explícita. "
            "Distinguir principio/regla_interpretativa/norma_organica/"
            "ficcion sin marcador textual requiere criterio humano."
        )

    has_enumeration, enumeration_closed = _detect_enumeration(text)

    structure = _derive_structure(statement_type, exception_present, has_enumeration)
    if structure is None and statement_type is not None:
        notes.append(
            f"Estructura: no determinada. El tipo de norma detectado "
            f"('{statement_type}') admite más de una estructura "
            f"compatible sin más evidencia "
            f"({sorted(COMPATIBILITY.get(statement_type, set()))})."
        )

    proleg_preview = None
    if exception_present:
        proleg_preview = build_proleg_preview(exception_marker, exception_scope, statement_type)

    return LabelProposal(
        statement_type=statement_type,
        structure=structure,
        deontic_modality=deontic_modality,
        hohfeldian_position=hohfeldian_position,
        addressee=addressee,
        antecedent_operator=antecedent_operator,
        exception_present=exception_present,
        exception_marker=exception_marker,
        exception_scope=exception_scope,
        presumption_rebuttable=presumption_rebuttable,
        notes=notes,
        proleg_preview=proleg_preview,
    )

File: latio-node/server/labeling/rules.mjs
// labeling/rules.mjs — port de src/labeling/rules.py.
//
// Motor de reglas deterministas para el etiquetado N3. Mismo contrato que el
// original: determined_fields (marcador léxico de baja ambigüedad),
// default_fields (sin marcador, valor modal documentado) y
// undetermined_fields (sin evidencia, nunca adivinado). No usa red, no usa
// LLM, no tiene costo — igual que la versión Python.

import * as lex from "./lexicalMarkers.mjs";
import { COMPATIBILITY } from "../models.mjs";
import { prove } from "../reasoning/engine.mjs";
import { Party, FactAction } from "../reasoning/models.mjs";
import { RuleBase, Rule, ExceptionLink, FactBase, FactEntry } from "../reasoning/models.mjs";

function detectAntecedentOperator(text) {
  for (const [label, pattern] of lex.ANTECEDENT_PATTERNS) {
    if (pattern.test(text)) return label;
  }
  return "ninguno_explicito";
}

function detectException(text) {
  for (const [markerLabel, pattern] of lex.EXCEPTION_PATTERNS) {
    const match = pattern.exec(text);
    if (match) {
      const tail = text.slice(match.index, match.index + 300);
      const scope = lex.ARTICLE_REFERENCE_RE.test(tail) ? "por_remision" : "interna";
      return { present: true, marker: markerLabel, scope };
    }
  }
  return { present: false, marker: null, scope: null };
}

function detectPresumption(text) {
  if (lex.PRESUMPTION_DE_DERECHO_RE.test(text)) return { isPresumption: true, rebuttable: false };
  if (lex.PRESUMPTION_LEGAL_RE.test(text)) return { isPresumption: true, rebuttable: true };
  return { isPresumption: false, rebuttable: null };
}

function detectRemision(text) {
  // Python: len(text.split()) — split() sin argumento colapsa cualquier
  // corrida de whitespace y descarta vacíos al inicio/fin. Equivalente en
  // JS: split en \s+ sobre el texto trimeado, filtrando vacío residual si
  // el texto es "".
  const trimmed = text.trim();
  const nWords = trimmed === "" ? 0 : trimmed.split(/\s+/).length;
  return nWords <= lex.REMISION_MAX_WORDS && lex.REMISION_LEAD_RE.test(text);
}

function detectDefinicion(text) {
  return lex.DEFINICION_LEAD_RE.test(text);
}

// Von Wright: obligación / prohibición / permiso. Orden importa — prohibición
// antes que permiso ("no podrá" no debe matchear "podrá").
function detectDeonticModality(text) {
  if (lex.DEONTIC_PROHIBICION_RE.test(text)) return "prohibicion";
  if (lex.DEONTIC_OBLIGACION_RE.test(text)) return "obligacion";
  if (lex.DEONTIC_PERMISO_RE.test(text)) return "permiso";
  return null;
}

// Correlato hohfeldiano por defecto de la modalidad deóntica detectada —
// simplificación declarada, no análisis bilateral completo de Hohfeld (no
// identifica contraparte). Ver docstring del original Python para la
// justificación completa de la bifurcación permiso->potestad/privilegio.
function deriveHohfeldFromDeontic(deonticModality, addressee) {
  if (deonticModality === "obligacion" || deonticModality === "prohibicion") return "deber";
  if (deonticModality === "permiso") {
    if (addressee === "juez" || addressee === "funcionario_o_notario") return "potestad";
    return "privilegio";
  }
  return "ninguno";
}

function detectAddressee(text) {
  if (lex.ADDRESSEE_JUEZ_RE.test(text)) return "juez";
  if (lex.ADDRESSEE_FUNCIONARIO_RE.test(text)) return "funcionario_o_notario";
  if (lex.ADDRESSEE_TERCERO_RE.test(text)) return "tercero";
  return null;
}

function detectEnumeration(text) {
  if (!lex.ENUMERATION_ITEM_RE.test(text)) return { has: false, closed: null };
  if (lex.ENUMERATION_CLOSED_RE.test(text)) return { has: true, closed: true };
  if (lex.ENUMERATION_OPEN_RE.test(text)) return { has: true, closed: false };
  return { has: true, closed: null }; // hay lista, abierta/cerrada sin determinar
}

// Usa COMPATIBILITY para derivar structure solo cuando queda un único
// candidato posible — nunca elige entre varios igual de válidos.
function deriveStructure(statementType, exceptionPresent, hasEnumeration) {
  if (statementType === null) return null;
  const candidates = COMPATIBILITY[statementType];
  if (!candidates || candidates.size === 0) return null;
  if (exceptionPresent) {
    return candidates.has("supuesto_consecuencia_con_excepcion")
      ? "supuesto_consecuencia_con_excepcion"
      : null;
  }
  let remaining = new Set(candidates);
  remaining.delete("supuesto_consecuencia_con_excepcion");
  if (!hasEnumeration) {
    remaining.delete("enumeracion_taxativa");
    remaining.delete("enumeracion_enunciativa");
  }
  if (remaining.size === 1) return [...remaining][0];
  return null;
}

// Instancia una RuleBase mínima y GENÉRICA a partir de la estructura ya
// detectada (regla + excepción) y corre el motor PROLEG real dos veces: sin
// la excepción probada, y con ella. No fabrica contenido semántico del
// artículo — ver docstring completo en el original Python (build_proleg_preview)
// para la distinción presunción/excepción sustantiva que documenta `note`.
export function buildProlegPreview(exceptionMarker, exceptionScope, statementType = null) {
  const rulebase = new RuleBase({
    id: "preview_derrotabilidad",
    description: `Estructura genérica derivada del artículo: regla con excepción ("${exceptionMarker}", scope=${exceptionScope}).`,
    rules: [
      new Rule({
        head: "consecuencia_aplica",
        body: ["antecedente_cumplido"],
        sourceNote: "Regla base detectada por reglas léxicas (no contenido inventado).",
      }),
    ],
    exceptions: [
      new ExceptionLink({ ruleHead: "consecuencia_aplica", exceptionHead: "excepcion_probada" }),
    ],
  });

  const baseFacts = new FactBase({
    entries: [
      new FactEntry({ action: FactAction.ALLEGE, fact: "antecedente_cumplido", party: Party.PLAINTIFF }),
      new FactEntry({ action: FactAction.PROVIDE_EVIDENCE, fact: "antecedente_cumplido", party: Party.PLAINTIFF }),
      new FactEntry({ action: FactAction.PLAUSIBLE, fact: "antecedente_cumplido", party: null }),
    ],
  });

  const withoutException = prove("consecuencia_aplica", Party.PLAINTIFF, rulebase, baseFacts);

  const withExceptionFacts = new FactBase({
    entries: [
      ...baseFacts.entries,
      new FactEntry({ action: FactAction.PLAUSIBLE, fact: "excepcion_probada", party: null }),
    ],
  });
  const withException = prove("consecuencia_aplica", Party.PLAINTIFF, rulebase, withExceptionFacts);

  let note;
  if (statementType === "presuncion") {
    note =
      "Vista previa estructural: usa el motor de razonamiento real sobre " +
      "una regla genérica con la misma forma detectada en el artículo. En " +
      "una presunción, esto modela la PRUEBA EN CONTRARIO (desplazamiento " +
      "de la carga de la prueba hacia quien quiere desvirtuarla), no la " +
      "derrota de una regla sustantiva — son figuras procesales distintas " +
      "aunque el motor las calcule con el mismo mecanismo. No es un " +
      "análisis semántico del contenido específico del artículo.";
  } else {
    note =
      "Vista previa estructural: usa el motor de razonamiento real sobre " +
      "una regla genérica con la misma forma detectada en el artículo " +
      "(regla + excepción sustantiva). No es un análisis semántico del " +
      "contenido específico del artículo — muestra que la excepción " +
      "detectada, si se prueba, efectivamente derrota la regla bajo el " +
      "motor determinista.";
  }

  return {
    rulebaseId: rulebase.id,
    withoutException: { proved: withoutException.proved, traceLength: withoutException.trace.length },
    withException: { proved: withException.proved, traceLength: withException.trace.length },
    note,
  };
}

export function proposeFromText(rawText) {
  // NFC: sin esto, el mismo texto en dos formas Unicode canónicamente
  // equivalentes (copiado desde macOS, o extraído de OCR, que suelen
  // producir NFD) puede no matchear los patrones léxicos (tildes
  // precompuestas) y dar un resultado distinto para el mismo enunciado.
  // text.normalize("NFC") de JS es equivalente a unicodedata.normalize de
  // Python — ambos implementan el mismo algoritmo Unicode estándar.
  const text = rawText.normalize("NFC").trim();

  const determined = [];
  const defaultFields = [];
  const undetermined = [];
  const notes = [];

  const antecedentOperator = detectAntecedentOperator(text);
  determined.push("antecedent_operator");

  const { present: exceptionPresent, marker: exceptionMarker, scope: exceptionScope } = detectException(text);
  determined.push("exception_present");
  if (exceptionPresent) {
    determined.push("exception_marker", "exception_scope");
  }

  let deonticModality = detectDeonticModality(text);
  const deonticDetermined = deonticModality !== null;
  if (deonticDetermined) {
    determined.push("deontic_modality");
  } else {
    deonticModality = "ninguno";
    undetermined.push("deontic_modality");
    notes.push(
      "Deóntica (Von Wright): sin marcador léxico de obligación/prohibición/" +
      "permiso — se deja 'ninguno' por defecto (correcto para definiciones/" +
      "remisiones, pero revisar si es una regla con deóntica implícita en " +
      "presente indicativo, ej. 'el comprador paga el precio' sin 'deberá')."
    );
  }

  // addressee se detecta ACÁ (antes de derivar Hohfeld) porque
  // deriveHohfeldFromDeontic necesita saber si el "permiso" se dirige a un
  // juez/funcionario (potestad) o a partes/tercero (privilegio).
  let addressee = detectAddressee(text);
  if (addressee !== null) {
    determined.push("addressee");
  } else if (deonticModality !== "ninguno") {
    addressee = "partes";
    defaultFields.push("addressee");
    notes.push(
      "Destinatario: sin marcador de juez/funcionario/tercero — se asume " +
      "'partes' por defecto (mayoría del derecho privado); revisar si el " +
      "artículo se dirige a otro destinatario."
    );
  } else {
    undetermined.push("addressee");
  }

  const hohfeldianPosition = deriveHohfeldFromDeontic(deonticDetermined ? deonticModality : null, addressee);
  if (deonticDetermined) {
    defaultFields.push("hohfeldian_position");
    notes.push(
      "Posición (Hohfeld): correlato por defecto de la deóntica detectada " +
      "(y, si es 'permiso', del destinatario) — no un análisis bilateral " +
      "de Hohfeld (no identifica contraparte) — ver " +
      "docs/limitaciones_conocidas.md §2."
    );
  } else {
    undetermined.push("hohfeldian_position");
  }

  let statementType = null;
  let statementTypeDetermined = false;
  let presumptionRebuttable = null;

  const { isPresumption, rebuttable } = detectPresumption(text);
  if (isPresumption) {
    statementType = "presuncion";
    presumptionRebuttable = rebuttable;
    statementTypeDetermined = true;
    determined.push("statement_type", "presumption_rebuttable");
  } else if (detectRemision(text)) {
    statementType = "remision";
    statementTypeDetermined = true;
    determined.push("statement_type");
  } else if (detectDefinicion(text)) {
    statementType = "definicion";
    statementTypeDetermined = true;
    determined.push("statement_type");
  } else if (deonticModality !== "ninguno") {
    // Hay marcador deóntico explícito y ningún otro marcador más específico:
    // el default modal es "regla".
    statementType = "regla";
    defaultFields.push("statement_type");
  } else {
    undetermined.push("statement_type");
    notes.push(
      "Tipo de norma: no determinado. Sin marcador léxico de presunción/" +
      "remisión/definición ni deóntica explícita. Distinguir " +
      "principio/regla_interpretativa/norma_organica/ficcion sin marcador " +
      "textual requiere criterio humano."
    );
  }

  const { has: hasEnumeration, closed: enumerationClosed } = detectEnumeration(text);

  const structure = deriveStructure(statementType, exceptionPresent, hasEnumeration);
  if (structure !== null) {
    (statementTypeDetermined ? determined : defaultFields).push("structure");
  } else {
    undetermined.push("structure");
    if (statementType !== null) {
      const candidates = [...(COMPATIBILITY[statementType] ?? new Set())].sort();
      notes.push(
        `Estructura: no determinada. El tipo de norma detectado ` +
        `('${statementType}') admite más de una estructura compatible sin ` +
        `más evidencia (${JSON.stringify(candidates)}).`
      );
    }
  }

  undetermined.push("generality");
  notes.push(
    "Generalidad (cantidad de condiciones, conceptos indeterminados, etc.) " +
    "queda fuera del alcance de este motor de reglas — completar manualmente."
  );

  let prolegPreview = null;
  if (exceptionPresent) {
    prolegPreview = buildProlegPreview(exceptionMarker, exceptionScope, statementType);
  }

  return {
    statementType,
    structure,
    deonticModality,
    hohfeldianPosition,
    addressee,
    antecedentOperator,
    exceptionPresent,
    exceptionMarker,
    exceptionScope,
    presumptionRebuttable,
    determinedFields: determined,
    defaultFields,
    undeterminedFields: undetermined,
    notes,
    prolegPreview,
    annotatedBy: "heuristica_local",
  };
}
